### DISCUSSION TITLE
How to pick point, cell, mesh with pyvista trame client mode?

### DISCUSSION THREAD

**Discussion by zhang9song** [pyvista/pyvista#5132] [Q&A]
Hi, 
Now, I can pick point, cell, mesh in pyvista trame server mode? How to pick point, cell, mesh with pyvista trame client mode? 
Thank you for your answer。

**Comment by banesullivan** [comment]
We haven't implemented this in our rendition of the trame viewer but it should be (is?) possible. @jourdain, would be the one know the most

**Comment by jourdain** [comment]
Yes, that can be supported but require some effort.

**Reply by zhang9song** [reply]
Thanks

**Comment by kerim371** [comment]
Hi,

Just wondering, any progress on this? I know picking is possible using Trame, but what about PyVista & Trame picking, is it possible now?

**Reply by jourdain** [reply]
The picking examples in [that directory](https://github.com/Kitware/trame-vtk/tree/master/examples/validation) illustrate how to do it, but they haven't been integrated into PyVista.

**Reply by kerim371** [reply]
Thank you!

Does that mean that picking doesn't work with `pyvista.Plotter` or with `pyvista.trame.ui.plotter_ui`?
For example I like how `pyvista.Plotter` looks like but it seems to be doesn't support picking. So should I avoid from using it if I need picking?

**Reply by banesullivan** [reply]
PyVista is just a wrapping around VTK so there's no reason you can't use PyVsita here. The following is modified version of the local picking example shared above with local rendering but using a PyVista plotter and our wrappers around trame-vtk's views.

```py
import numpy as np
import pyvista as pv
from pyvista.trame.ui import plotter_ui
from trame.app import get_server
from trame.ui.html import DivLayout
from trame.widgets import client, html


server = get_server(client_type="vue3")
ctrl = server.controller

plotter = pv.Plotter(off_screen=True)
cone_actor = plotter.add_mesh(pv.Cone(), color="r", opacity=0.75)
sphere_actor = plotter.add_mesh(pv.Sphere(), color="b", opacity=0.75)


def on_click(event):
    if event is None:
        return
    point = event.get("worldPosition")
    # Why does local rendering not respect the point size occasionally?
    plotter.add_points(
        np.array(point), color="pink", point_size=20, name="picked_point"
    )


with DivLayout(server) as layout:
    client.Style("html, body { padding: 0; margin: 0; }")
    layout.root.style = "width: 100vw; height: 100vh;"

    with plotter_ui(
        plotter,
        mode="client",  # logic above assumes client/local rendering
        # NOTE: if picking with remote rendering then just use PyVista's robust picking interfaces
        picking_modes=("picking_modes", ["click"]),
        click=(on_click, "[$event]"),
    ) as view:
        ctrl.get_scene_object_id = view.get_scene_object_id


server.start()
```

**Reply by kerim371** [reply]
Oh, thank you very much for that example!
That helps me a lot!
